### PR TITLE
Improve cellPngDec

### LINKED ISSUE #2028: HLE: cellpngdec decodes wrong dimension
During workaround the span error found in Sonic CD , we found wrong dimension got decoded from HLE Cellpngdec for the following frame that trigger the span error.

(Actually it happened long time ago before the rewrite of pngdec library that have this texture distored)

HLE

LLE

### PR DESCRIPTION
**Pull request by jarveson**
- ReadHeader and SetParam corrected
- Partial loading Implemented
- Interlace Support added
- Improve error handling

This fixes #2028 and now should report back correct sizes from their respective calls.
Tested with Sonic CD and Arkedo
Partial and interlace tested with pngdec_ppu*.elfs

It *should* take care of #1871, but nothing is displayed with LLE pngdec also, so its unknown

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellPngDec.cpp [hunk 1]
```diff
@@ -401,43 +464,43 @@ s32 pngDecOpen(ppu_thread& ppu, PHandle handle, PPStream png_stream, PSrc source
...
-	stream->info.colorSpace = getPngDecColourType(png_get_color_type(stream->png_ptr, stream->info_ptr));
-	stream->info.bitDepth = png_get_bit_depth(stream->png_ptr, stream->info_ptr);
-	stream->info.interlaceMethod = png_get_interlace_type(stream->png_ptr, stream->info_ptr);
-	stream->info.chunkInformation = pngDecGetChunkInformation(stream);
+		// push header tag to libpng to keep us in sync
+		try 
+		{
+			png_process_data(stream->png_ptr, stream->info_ptr, header, 8);
+		}
+		catch (...){
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: What it's supposed to catch from C function? Also formatting.

Comment (paulsapps) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: In win32 with certain build options this a hack to catch access violations from reading/writing invalid memory and things... definitely not recommended!

Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: I'm curious about setjmp removed.

Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: Regarding access violations, PS3 memory shouldn't be passed to such functions in the first place.

Comment (jarveson) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: ok, its not a hack. It's a weirdness in the libpng api. Theres two ways to set it up, either with the setjmp or with using your own error_fn, which we already had setup, https://github.com/RPCS3/rpcs3/pull/2394/files/b98aef0f7931dd69890436ae2abd0828924f05b0#diff-c5a75817f68ae08ed5af8e372459ebc2L87

The plan wasn't to 'hack and catch access violations'. I can make it throw / catch a specific exception if that helps?

Comment (paulsapps) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: Why not catch that specific exception? catch(...) can also catch int* ptr =0; *ptr = 7; depending on the compiler options which is why it should usually never be used.

File: rpcs3/Emu/Cell/Modules/cellPngDec.cpp [hunk 2]
```diff
@@ -742,13 +840,51 @@ s32 cellPngDecClose(ppu_thread& ppu, PHandle handle, PStream stream)
...
+		extInfo->reserved = 0;
+	}
+
+	// lets push what we have so far
+	u8* data = static_cast<u8*>(stream->buffer->data.get_ptr()) + stream->buffer->cursor;
+	try 
+	{
+		png_process_data(stream->png_ptr, stream->info_ptr, data, stream->buffer->length);
+	}
+	catch (LibPngCustomException)
```
Comment (danilaml) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: why not catch by (const) ref?

**Comment by Nekotekina** [comment]
C functions shouldn't propagate any exceptions... I have no idea how it should be implemented.

**Comment by jarveson** [comment]
Why not? It's still our exception, unless I'm missing something about mixing c and exceptions.

I think i can switch it all over to the setjmp method, but then its just a glorifed goto at the top of every function. 

It seems like its going to be a bit weird either way?

File: rpcs3/Emu/Cell/Modules/cellPngDec.cpp (hunk 2 above)
Comment (jarveson) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: because theres no variable?

Comment (danilaml) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: @jarveson and? You are still catching by value. It'd make a copy. Take a look [here](http://stackoverflow.com/questions/7401269/catch-by-reference-when-exception-variable-is-not-defined).

**Comment by danilaml** [comment]
@Nekotekina I saw few mentions that you may indeed throw if you want from png error handler (it's not designed to return) but the official manual is unclear (it just mentions C++ throw/catch methods, but in a weird way): http://www.libpng.org/pub/png/libpng-1.2.5-manual.html#section-5.1

File: rpcs3/Emu/Cell/Modules/cellPngDec.cpp (hunk 2 above)
Comment (paulsapps) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: const LibPngCustomException& is the usual way

Comment (danilaml) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: The SO answer suggest that `LibPngCustomException&` would be the safest (and shortest) way)

Comment (jarveson) [inline rpcs3/Emu/Cell/Modules/cellPngDec.cpp]: Weird, ill throw some '&'s in. 😃

**Pull request merged**